### ISSUE TITLE
tags and branches??

### ISSUE THREAD

**Issue by l0k1verloren** [therecipe/qt#990]
I may be out of date a few weeks but last time I looked at the tree of this had a master branch, no tags, and at the time had already nodes above the master branch when I cloned it.

Go modules really require version tags, and at minimum there should be a tag representing the latest current version. There really should be more than just a master branch also, although the repo is fairly stable.

One bad branch on the master and a lot of sad gophers will be made.

**Comment by therecipe** [comment]
Hey

Yeah, I wrote a bit about my reasoning for the current state of non versioning here https://github.com/therecipe/qt/issues/973#issuecomment-538797814
The summary is though, that I can't really enforce pinning of Go nor Qt and so providing versions/tags would only lead to a false sense of security I think.
(And if you already pin Go or Qt yourself, then pinning the binding is not that much of additional work either)

I might start versioning, if there are mayor breaking changes for which I can't provide backward compatibility measures though, but there was simply no need for this yet.
Also pseudo versions seem to work good enough for go modules, so I don't really feel the need to provide tags.

However, this is not set in stone and if there are reasons to provide versioning then I'm open for them as well. It's just that I can't see a clear benefit of doing so at the moment.

**Comment by l0k1verloren** [comment]
My primary concern was just tags for module versioning. It seems to behave ok currently but I have found with several of my own repositories that modules behave unpredictably in the absence of tags. At least to fix the version that will be pulled by `go mod download` or similar so if you accidentally make a push that breaks something your environment masked or so and it breaks apps that import it... 

It would be cool if there was something that could short cut and cache the binary objects from Qt that are needed in the build so they didn't need any recompilation (and qt doesn't seem to use much threads to compile and I can't see how to provoke it to parallel build), but this issue here just a small insurance against an accidental bad commit, putting a version tag on the latest good commit and moving it up lagging behind progress so there is time for bugs to be found and squashed before advancing it.

**Comment by therecipe** [comment]
Yeah, I thought about a moving tag as well, but it sadly won't work because Go hashes and compares the content of the modules.
But `go mod download` and `go install -v -tags=no_env github.com/therecie/qt/cmd/...` should only pull/install the version pinned in your go.mod file though. (`go get ...` on the other side will always attempt to pull the latest master)

If you need extra security, then I would advice you to just fork and tag the repo and then use the "replace" directive inside the go.mod file. That way your code would continue to work even if this repo gets deleted or something.

About the caching of the compilation, yeah I hope a new/better linker (http://golang.org/s/better-linker) will speedup the whole process in the future. But in the meantime you can speedup the process on linux by using ccache or the gold linker: https://github.com/therecipe/qt/wiki/Faster-builds-(Linux)

**Comment by l0k1verloren** [comment]
I don't think it matters if the hash is different if you specify only the version in the go.mod but to be honest I'm a bit cheesed with modules anyway, it makes my build environment more brittle.

**Comment by therecipe** [comment]
Ah, yes you are right, this could work if you don't commit the `go.sum` file or at least remove the `therecipe/qt` entries from it.
I will think about it.

And I also had another thought about versioning in general, and it could be that I changed my mind about it.

Btw, I don't really like the way go modules work at the moment as well.
The whole thing is just too complex and brittle.
